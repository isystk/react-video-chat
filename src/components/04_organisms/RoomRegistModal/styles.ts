import { makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles((theme) => ({
  form: {},
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
}))

export { useStyles }
