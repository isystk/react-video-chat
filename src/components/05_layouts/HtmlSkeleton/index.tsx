import React, { Children, FC, useState } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import useAppRoot from '@/stores/useAppRoot'
import Header from '@/components/04_organisms/Header'
import SideMenu from '@/components/04_organisms/SideMenu'
import Title, { TitleProps } from './Title'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { CssBaseline } from '@material-ui/core'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: TitleProps['children']
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  main,
  title,
  description,
  isMenuOpen,
  setMenuOpen,
  children,
  ...props
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
    </Head>
    <Context.Provider value={main}>
      <CssBaseline />
      <Header isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} />
      <div>{children}</div>
      <SideMenu isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} />
    </Context.Provider>
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  const main = useAppRoot()
  const [isMenuOpen, setMenuOpen] = useState(false)

  if (!main) return <></>

  const APP_NAME = process.env.APP_NAME

  children = Children.map(children, (child) =>
    isReactElement(child) && child.type === Title
      ? (title = `${child.props.children} | ${APP_NAME}`) && undefined
      : React.cloneElement(child)
  )
  if (!title) title = process.env.APP_NAME

  const description = `${APP_NAME} is an application that allows you to casually chat with your friends!`
  return presenter({
    main,
    title,
    description,
    isMenuOpen,
    setMenuOpen,
    children,
    ...props,
  })
}

export const Context = React.createContext(null)

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
