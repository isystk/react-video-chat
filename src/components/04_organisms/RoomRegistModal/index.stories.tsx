import { Meta, Story } from '@storybook/react'
import React from 'react'
import RoomRegistModal from './index'
import MainService from '@/services/main'
import { Context } from '@/components/05_layouts/HtmlSkeleton'

export default {
  title: '04_organisms/RoomRegistModal',
  component: RoomRegistModal,
} as Meta

const Template: Story = (props) => {
  const main = new MainService(() => ({}))
  main.room.isOpen = true
  return (
    <Context.Provider value={main}>
      <RoomRegistModal />
    </Context.Provider>
  )
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
