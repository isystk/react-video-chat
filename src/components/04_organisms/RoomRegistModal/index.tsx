import React, { FC, useContext } from 'react'
import * as Yup from 'yup'
import Modal from '@/components/01_atoms/Modal'
import { ContainerProps, WithChildren } from 'types'
import * as styles from './styles'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import { connect } from '@/components/hoc'
import MainService from '@/services/main'
import { Room } from '@/services/room'
import {
  Button,
  Grid,
  Typography,
  Paper,
  TextField,
  Fade,
  FormGroup,
  Container,
} from '@mui/material'
import { Formik, Form } from 'formik'
import AddCircleIcon from '@mui/icons-material/AddCircle'

/** RoomRegistModalProps Props */
export type RoomRegistModalProps = WithChildren & { room: Room }
/** Presenter Props */
export type PresenterProps = RoomRegistModalProps & {
  main
  videoInputDefault
  changeDevice
  videoInput
  audioInputDefault
  audioInput
  audioOutputDefault
  audioOutput
}

/** Presenter Component */
const RoomRegistModalPresenter: FC<PresenterProps> = ({
  main,
  initialValues,
  validationSchema,
  onSubmit,
  onClose,
  ...props
}) => (
  <>
    <Modal isOpen={main.room.isOpen} handleClose={onClose}>
      <Container component="main" maxWidth="xs">
        <Fade in={main.room.isOpen}>
          <Paper style={{ marginTop: '25px' }}>
            <Formik
              initialValues={initialValues}
              validationSchema={validationSchema}
              onSubmit={onSubmit}
            >
              {({ values, errors, touched, handleChange, handleBlur }) => (
                <Form>
                  <Grid
                    container
                    spacing={3}
                    style={{ padding: '20px' }}
                    justifyContent="center"
                  >
                    <Grid item xs={12} sm={12} md={12}>
                      <FormGroup style={{ margin: 8 }}>
                        <Typography style={{ marginLeft: 8 }}>
                          部屋名
                        </Typography>
                        <TextField
                          fullWidth
                          margin="normal"
                          variant="outlined"
                          className="red-border"
                          type="text"
                          name="name"
                          value={values.name}
                          onChange={handleChange}
                          onBlur={handleBlur}
                          error={!!(touched.name && errors.name)}
                          helperText={errors.name}
                        />
                        <Typography style={{ marginLeft: 8 }}>本文</Typography>
                        <TextField
                          fullWidth
                          margin="normal"
                          multiline
                          rows={4}
                          InputLabelProps={{ shrink: true }}
                          variant="outlined"
                          type="text"
                          name="description"
                          value={values.description}
                          onChange={handleChange}
                          onBlur={handleBlur}
                          error={!!(touched.description && errors.description)}
                          helperText={errors.description}
                        />
                        <Grid
                          container
                          style={{ paddingTop: 30 }}
                          justify="center"
                          direction="row"
                        >
                          <Grid item>
                            <Button
                              variant="contained"
                              color="primary"
                              type="submit"
                              startIcon={<AddCircleIcon />}
                            >
                              登録する
                            </Button>
                          </Grid>
                        </Grid>
                      </FormGroup>
                    </Grid>
                  </Grid>
                </Form>
              )}
            </Formik>
          </Paper>
        </Fade>
      </Container>
    </Modal>
  </>
)

/** Container Component */
const RoomRegistModalContainer: React.FC<
  ContainerProps<RoomRegistModalProps, PresenterProps>
> = ({ presenter, children, room, ...props }) => {
  const main = useContext<MainService | null>(Context)
  if (!main) return <></>

  const isNew = !room

  // フォームの初期値
  let initialValues = {
    name: '',
    description: '',
  }
  if (room) {
    initialValues = {
      name: room.name ?? '',
      description: room.description ?? '',
    }
  }
  // フォームのバリデーション
  const validationSchema = Yup.object().shape({
    name: Yup.string().required('部屋名を入力してください。'),
    description: Yup.string().required('説明文を入力してください。'),
  })

  const onClose = () => {
    main.room.isOpen = false
    main.setAppRoot()
  }

  // フォームの送信
  const onSubmit = async (values: Form) => {
    console.log(values)
    if (isNew) {
      const newRoom = { ...values } as Room
      await main?.room.createRoom(newRoom)
    } else {
      const newRoom = { ...values, id: room.id } as Room
      await main?.room.updateRoom(newRoom)
    }
    onClose()
  }

  return presenter({
    children,
    main,
    initialValues,
    validationSchema,
    onSubmit,
    onClose,
    ...props,
  })
}

export default connect<RoomRegistModalProps, PresenterProps>(
  'RoomRegistModal',
  RoomRegistModalPresenter,
  RoomRegistModalContainer
)
