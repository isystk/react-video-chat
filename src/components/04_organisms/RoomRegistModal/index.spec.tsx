import React from 'react'
import renderer from 'react-test-renderer'
import RoomRegistModal from './index'
import '@testing-library/jest-dom/extend-expect'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import MainService from '@/services/main'

describe('RoomRegistModal', () => {
  it('Match Snapshot', () => {
    const main = new MainService(() => ({}))
    main.mediaDevice.isOpen = true
    const component = renderer.create(
      <Context.Provider value={main}>
        <RoomRegistModal />
      </Context.Provider>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
