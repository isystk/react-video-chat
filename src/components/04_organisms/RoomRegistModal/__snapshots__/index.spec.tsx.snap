// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomRegistModal Match Snapshot 1`] = `null`;
